File: Algorithm/Other/leetcode/14. Longest Common Prefix/LongestCommonPrefix.ts
// 以下はTypeScriptで実装した**最長共通接頭辞 (Longest Common Prefix)** の関数です。

// ### 特徴

// * `fs` モジュールで**処理時間**と**メモリ消費量**を`log.txt`に出力します。
// * 処理は関数ベースで実装。
// * パラメータと戻り値はすべてコメント付き。

// ---

// ```typescript
import * as fs from 'fs';

/**
 * 最長共通接頭辞を求める関数
 * @param strs - string型の配列（最長共通接頭辞を求める対象）
 * @returns string - 最長共通接頭辞。共通部分がなければ空文字を返す。
 */
function longestCommonPrefix(strs: string[]): string {
    if (strs.length === 0) return "";

    let prefix = strs[0];

    for (let i = 1; i < strs.length; i++) {
        // prefixが一致しなくなるまで短くしていく
        while (strs[i].indexOf(prefix) !== 0) {
            prefix = prefix.substring(0, prefix.length - 1);
            if (prefix === "") return "";
        }
    }

    return prefix;
}

/**
 * パフォーマンス計測と結果ログ出力用の関数
 * @param strs - string型の配列（最長共通接頭辞を求める対象）
 * @param outputPath - ログファイルの出力先パス
 */
function runWithPerformanceLog(strs: string[], outputPath: string): void {
    const startTime = process.hrtime.bigint(); // 高精度時間計測
    const startMem = process.memoryUsage().heapUsed;

    const result = longestCommonPrefix(strs);

    const endTime = process.hrtime.bigint();
    const endMem = process.memoryUsage().heapUsed;

    const timeElapsedMs = Number(endTime - startTime) / 1_000_000; // ミリ秒に変換
    const memoryUsageKb = (endMem - startMem) / 1024; // KBに変換

    const logData = `
==== Longest Common Prefix Execution ====
Input: ${JSON.stringify(strs)}
Output: ${result}
Time Elapsed: ${timeElapsedMs.toFixed(3)} ms
Memory Usage: ${memoryUsageKb.toFixed(3)} KB
=========================================
`;

    fs.writeFileSync(outputPath, logData);
}

// サンプル実行
const example1 = ["flower", "flow", "flight"];
const example2 = ["dog", "racecar", "car"];

runWithPerformanceLog(example1, 'log_example1.txt');
runWithPerformanceLog(example2, 'log_example2.txt');
// ```

// ---

// ## 実行環境

// * TypeScript 5.1
// * Node.js 18.16.1

// ## 出力ファイル

// * `log_example1.txt`
// * `log_example2.txt`

// それぞれに**処理時間**と**メモリ消費量**が記録されます。

// ---

// ## 補足

// * `process.hrtime.bigint()` を使うことでナノ秒精度の時間計測ができます。
// * `process.memoryUsage().heapUsed` はNode.jsのヒープ領域の消費メモリ（バイト）を返します。
// * `fs.writeFileSync` を使うことで、非同期制御なしで確実にログが出力されます。

File: Algorithm/Other/leetcode/14. Longest Common Prefix/LongestCommonPrefix.py
# LeetCode のテストケースは
# `class Solution` にメソッドを定義する形式が必要です。

# なので、次のように修正します：

# ## ✅ LeetCode形式（`class Solution`）

# ```python
import time
import tracemalloc
from typing import List, Tuple

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        """
        LeetCode用 最長共通接頭辞関数

        :param strs: List[str] - 文字列のリスト
        :return: str - 最長共通接頭辞。共通部分がなければ空文字。
        """
        if not strs:
            return ""
        
        prefix = strs[0]
        
        for s in strs[1:]:
            while not s.startswith(prefix):
                prefix = prefix[:-1]
                if not prefix:
                    return ""
        return prefix

def run_with_performance_log(strs: List[str]) -> Tuple[str, float, int]:
    """
    処理時間とメモリ消費量を計測し、最長共通接頭辞を求める

    :param strs: List[str] - 文字列リスト
    :return: Tuple[str, float, int]
        - 最長共通接頭辞（str）
        - 処理時間（ミリ秒、float）
        - メモリ消費量（KB、int）
    """
    solution = Solution()

    tracemalloc.start()
    start_time = time.perf_counter()

    result = solution.longestCommonPrefix(strs)

    end_time = time.perf_counter()
    _, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    time_elapsed_ms = (end_time - start_time) * 1000  # ミリ秒
    memory_usage_kb = peak // 1024  # KB単位に変換

    return result, time_elapsed_ms, memory_usage_kb

# サンプル実行
if __name__ == "__main__":
    test_cases = [
        ["flower", "flow", "flight"],
        ["dog", "racecar", "car"]
    ]

    for i, case in enumerate(test_cases, 1):
        result, time_ms, mem_kb = run_with_performance_log(case)
        print(f"Test Case {i}: {case}")
        print(f"Longest Common Prefix: '{result}'")
        print(f"Time: {time_ms:.3f} ms")
        print(f"Memory: {mem_kb} KB")
        print("=" * 40)
# ```

# ---

# ## ✅ 実行結果（例）

# ```
# Test Case 1: ['flower', 'flow', 'flight']
# Longest Common Prefix: 'fl'
# Time: 0.030 ms
# Memory: 0 KB
# ========================================
# Test Case 2: ['dog', 'racecar', 'car']
# Longest Common Prefix: ''
# Time: 0.025 ms
# Memory: 0 KB
# ========================================
# ```

# ---

# ## LeetCode用注意点：

# * `Solution`クラスにメソッドを書く必要があります。
# * 引数や戻り値は LeetCode標準形式。

# ---
